crates/compiler/constrain/src/expr.rs

```diff
@@ -134,7 +135,12 @@ fn constrain_untyped_closure(
     vars.push(closure_var);
     vars.push(fn_var);
 
-    let body_type = constraints.push_expected_type(NoExpectation(return_type_index));
+    let body_type = constraints.push_expected_type(ForReason(
+        Reason::Return,
```

Comment:
Not sure Reason::Return here is appropriate, because there is no `return` keyword for the value of function body.